packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -688,154 +861,585 @@ class _SortData {
-      // leftmost one.
-      final List<_SortData> inBandOfTop = inBand(topmost, candidates).toList();
-      inBandOfTop.sort(compareBeginningSide);
-      if (inBandOfTop.isNotEmpty) {
-        return inBandOfTop.first;
+      // Now that we know there are others in the same band as the topmost, then pick
+      // the one at the beginning, depending on the text direction in force.
+
+      // Find out the directionality of the nearest common Directionality
+      // ancestor for all nodes. This provides a base directionality to use for
+      // the ordering of the groups.
+      final TextDirection nearestCommonDirectionality = _ReadingOrderSortData.commonDirectionalityOf(inBandOfTop);
+
+      // Do an initial common-directionality-based sort to get consistent geometric
+      // ordering for grouping into directionality groups. It has to use the
+      // common directionality to be able to group into sane groups for the
+      // given directionality, since rectangles can overlap and give different
+      // results for different directionalities.
+      mergeSort<_ReadingOrderSortData>(inBandOfTop, compare: (_ReadingOrderSortData a,_ReadingOrderSortData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+
+      // Collect the top band into internally sorted groups with shared directionality.
+      final List<_ReadingOrderDirectionalGroupData> bandGroups = collectDirectionalityGroups(inBandOfTop);
+      if (bandGroups.length == 1) {
+        // There's only one directionality group, so just send back the first
+        // one in that group, since it's already sorted.
+        return bandGroups.first.members.first;
       }
-      return topmost;
+
+      // Sort the groups based on the common directionality and bounding boxes.
+      mergeSort<_ReadingOrderDirectionalGroupData>(bandGroups, compare: (_ReadingOrderDirectionalGroupData a, _ReadingOrderDirectionalGroupData b) {
+        switch(nearestCommonDirectionality) {
+          case TextDirection.ltr:
+            return a.rect.left.compareTo(b.rect.left);
+          case TextDirection.rtl:
+            return b.rect.right.compareTo(a.rect.right);
+        }
+        assert(false, 'Unhandled directionality $nearestCommonDirectionality');
+        return 0;
+      });
+      return bandGroups.first.members.first;
     }
 
-    final List<_SortData> data = <_SortData>[
-      for (final FocusNode node in nodes) _SortData(node),
+    final List<_ReadingOrderSortData> data = <_ReadingOrderSortData>[
+      for (final FocusNode node in descendants) _ReadingOrderSortData(node),
     ];
 
     // Pick the initial widget as the one that is leftmost in the band of the
```

Comment:
Is this comment ("leftmost") still correct? Doesn't that depend on directionality?

Reply:
Fixed.